### PR TITLE
update dfs

### PR DESCRIPTION
**Pull request by geniusgogo**
## 拉取/合并请求描述：(PR description)

[
     累积修复dfs的改动，涉及到强制卸载、pread/pwrite等改动
]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

**Comment by BernardXiong** [comment]
```bash
/home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c: In function 'dfs_device_fs_stat':
/home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c:289:18: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
  289 |     st->st_dev = (dev_t)dfs_filesystem_lookup(fs->path);
      |                  ^
```

**Comment by geniusgogo** [comment]
> ```shell
> /home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c: In function 'dfs_device_fs_stat':
> /home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c:289:18: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
>   289 |     st->st_dev = (dev_t)dfs_filesystem_lookup(fs->path);
>       |                  ^
> ```

本地验证：gcc12.0.1 没有遇到这个问题。是不是CI的编译器有点旧？

File: components/dfs/dfs_v2/filesystems/elmfat/dfs_elm.c [hunk 1]
```diff
@@ -507,15 +507,13 @@ int dfs_elm_close(struct dfs_file *file)
     else if (file->vnode->type == FT_REGULAR)
     {
         FIL *fd = RT_NULL;
+
         fd = (FIL *)(file->vnode->data);
         RT_ASSERT(fd != RT_NULL);
 
-        result = f_close(fd);
```
Comment (BernardXiong) [inline components/dfs/dfs_v2/filesystems/elmfat/dfs_elm.c:513]: 为什么反而把错误检查忽略了？

File: components/dfs/dfs_v2/src/dfs_file.c [hunk 2]
```diff
@@ -48,11 +48,28 @@ ssize_t rw_verify_area(struct dfs_file *file, off_t *ppos, size_t count)
...
 {
     if (file)
     {
-        rt_mutex_take(&file->pos_lock, RT_WAITING_FOREVER);
+        if (file->vnode->type != FT_REGULAR)
+        {
+            rt_mutex_take(&file->pos_lock, RT_WAITING_FOREVER);
+        }
         file->fpos = fpos;
         rt_mutex_release(&file->pos_lock);
```
Comment (BernardXiong) [inline components/dfs/dfs_v2/src/dfs_file.c:74]: 前面有判断，后面没判断，会出问题吧。

File: components/dfs/dfs_v2/src/dfs_file.c [hunk 3]
```diff
@@ -361,10 +389,8 @@ int dfs_file_open(struct dfs_file *file, const char *path, int oflags, mode_t mo
 
                         /* follow symbol link */
                         target_dentry = dfs_file_follow_link(dentry);
-                        if (target_dentry)
-                        {
-                            dentry = target_dentry;
-                        }
+                        dfs_dentry_unref(dentry);
```
Comment (BernardXiong) [inline components/dfs/dfs_v2/src/dfs_file.c:392]: 是否依然需要对返回值进行判断？

**Review by BernardXiong — CHANGES_REQUESTED** [review]

File: components/dfs/dfs_v2/filesystems/elmfat/dfs_elm.c (hunk 1 above)
Comment (geniusgogo) [reply components/dfs/dfs_v2/filesystems/elmfat/dfs_elm.c:513]: 在强制卸载的情况下检查返回值已经没有意义了，即使返回出错也必须释放掉。

File: components/dfs/dfs_v2/src/dfs_file.c (hunk 2 above)
Comment (geniusgogo) [reply components/dfs/dfs_v2/src/dfs_file.c:74]: 当前的语义环境下，dfs_file_set_fpos和 dfs_file_get_fpos是成对出现的，一个上锁，一个下锁，上锁也做了判断，下锁也做了判断。

这里严格来讲后续是不需要区分目录还是文件的。统一上锁才是。

File: components/dfs/dfs_v2/src/dfs_file.c (hunk 3 above)
Comment (geniusgogo) [reply components/dfs/dfs_v2/src/dfs_file.c:392]: dentry中途不会修改的，上面if已经判断过了。

**Review by BernardXiong — APPROVED** [review]

**Comment by meng-plus** [comment]
> ```shell
> /home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c: In function 'dfs_device_fs_stat':
> /home/runner/work/rt-thread/rt-thread/components/dfs/dfs_v1/filesystems/devfs/devfs.c:289:18: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
>   289 |     st->st_dev = (dev_t)dfs_filesystem_lookup(fs->path);
>       |                  ^
> ```


iar_９环境下无法编译通过，建议修改未
    st->st_dev = (rt_device_t)dfs_filesystem_lookup(fs->path);

**Pull request merged**